lib/spack/llnl/util/filesystem.py

```diff
@@ -472,3 +474,110 @@ def find_library_path(libname, *paths):
         if os.path.exists(library):
             return path
     return None
+
+# Utilities for libraries
+
+
+class LibraryList(collections.Sequence):
+    """Manages a list of absolute paths to libraries"""
+    def __init__(self, libraries):
+        self.libraries = list(libraries)
+
+    @property
+    def directories(self):
+        return [os.path.dirname(x) for x in self.libraries]
+
+    @property
+    def libnames(self):
+        return [os.path.basename(x).split('.')[0] for x in self.libraries]
+
+    @property
+    def names(self):
+        return [x[3:] for x in self.libnames]
+
+    @property
+    def search_flags(self):
+        return ['-L' + x for x in set(self.directories)]
+
+    @property
+    def link_flags(self):
+        return ['-l' + name for name in self.names]
+
+    def __getitem__(self, item):
+        cls = type(self)
+        if isinstance(item, numbers.Integral):
+            return self.libraries[item]
+        return cls(self.libraries[item])
+
+    def __eq__(self, other):
+        return self.libraries == other.libraries
+
+    def __len__(self):
+        return len(self.libraries)
+
+    def joined(self, separator=' '):
+        return separator.join(self.libraries)
+
+    def __repr__(self):
+        return self.__class__.__name__ + '(' + repr(self.libraries) + ')'
+
+    def __str__(self):
+        return self.joined()
+
+
+def find_libraries(args, root, shared=True, recurse=False):
+    """Returns an iterable object containing a list of full paths to
+    libraries if found.
+
+    Args:
+        args: iterable object containing a list of library names to
+            search for (e.g. 'libhdf5')
+        root: root folder where to start searching
+        shared: if True searches for shared libraries, otherwise for static
+        recurse: if False search only root folder, if True descends top-down
+            from the root
+
+    Returns:
+        list of full paths to the libraries that have been found
+    """
+    # Construct the right suffix for the library
+    if shared is True:
+        suffix = 'dylib' if sys.platform == 'darwin' else 'so'
+    else:
+        suffix = 'a'
+    # List of libraries we are searching with suffixes
+    libraries = ['{0}.{1}'.format(lib, suffix) for lib in args]
+    # Search method
+    if recurse is False:
+        search_method = _find_libraries_non_recursive
+    else:
+        search_method = _find_libraries_recursive
+
+    return search_method(libraries, root)
+
+
+def _find_libraries_recursive(libraries, root):
```

Comment:
maybe hide this one and next from from users inside `class _Internal:`, see http://stackoverflow.com/a/551048

i.e. make sure that inside packages only `find_libraries(args, root, shared=True, recurse=False)` can be called and the two functions are for internal usage only (similar to `private` for classes in C++ or Unnamed namespaces for utility functions)